packages/custom-webpack/src/custom-webpack-builder.ts

```diff
@@ -1,25 +1,61 @@
-import {CustomWebpackBuilderConfig} from "./custom-webpack-builder-config";
-import {Configuration} from "webpack";
-import {getSystemPath, Path} from '@angular-devkit/core';
-import {mergeConfigs} from "./webpack-config-merger";
+import { getSystemPath, Path } from '@angular-devkit/core';
+import { Configuration } from 'webpack';
+
+import { mergeConfigs } from './webpack-config-merger';
+import { CustomWebpackBuilderConfig } from './custom-webpack-builder-config';
 
 export const defaultWebpackConfigPath = 'webpack.config.js';
 
+type CustomWebpackConfig =
+    | Configuration
+    | Promise<Configuration>
+    | ((baseWebpackConfig: Configuration, buildOptions: any) => Configuration)
+    | ((baseWebpackConfig: Configuration, buildOptions: any) => Promise<Configuration>);
+
 export class CustomWebpackBuilder {
-    static buildWebpackConfig(root: Path,
-      config: CustomWebpackBuilderConfig,
-      baseWebpackConfig: Configuration,
-      buildOptions: any): Configuration {
+    static buildWebpackConfig(
+        root: Path,
+        config: CustomWebpackBuilderConfig,
+        baseWebpackConfig: Configuration,
+        buildOptions: any
+    ): Configuration | Promise<Configuration> {
         if (!config) {
             return baseWebpackConfig;
         }
+
         const webpackConfigPath = config.path || defaultWebpackConfigPath;
-        const customWebpackConfig = require(`${getSystemPath(root)}/${webpackConfigPath}`);
-        //TODO: support function returning promise
-        if (typeof customWebpackConfig === "function") {
-            return customWebpackConfig(baseWebpackConfig, buildOptions);
-        } else {
-            return mergeConfigs(baseWebpackConfig, customWebpackConfig, config.mergeStrategies, config.replaceDuplicatePlugins);
+        const configOrFactoryOrPromise: CustomWebpackConfig = require(`${getSystemPath(
+            root
+        )}/${webpackConfigPath}`);
+
+        if (configOrFactoryOrPromise instanceof Promise) {
```

Comment:
I guess it's a matter of personal preference but I prefer the `async/await` syntax. Since you can await on non-promises (which just creates a promise of awaited object) the code becomes cleaner and shorter:
```typescript
async buildWebpackConfig(
        root: Path,
        config: CustomWebpackBuilderConfig,
        baseWebpackConfig: Configuration,
        buildOptions: any
    ): Promise<Configuration> {
        if (!config) {
            return baseWebpackConfig;
        }

        const webpackConfigPath = config.path || defaultWebpackConfigPath;
        const configOrFactoryOrPromise: CustomWebpackConfig = require(`${getSystemPath(
            root
        )}/${webpackConfigPath}`);

        if (typeof configOrFactoryOrPromise === 'function') {
            // That exported function can be synchronous either
            // asynchronous. Given the following example:
            // `module.exports = async (config) => { ... }`
            return configOrFactoryOrPromise(baseWebpackConfig, buildOptions);
        }

        const resolvedConfig = await configOrFactoryOrPromise;
        // In this case the user has exported a plain object, like:
        // `module.exports = { ... }`
        return mergeConfigs(
            baseWebpackConfig,
            resolvedConfig,
            config.mergeStrategies,
            config.replaceDuplicatePlugins
        );
    }
```

Reply:
I thought about this, but will need to fix all tests since the Webpack config will be always “gotten” asynchronously.